src/rpc/client.cpp

```diff
@@ -6,195 +6,253 @@
+                        err.requiredParams,
+                        err.requiredParams + err.optionalParams,
+                        err.providedArgs);
+            },
+            [](const UnparseableArgument& err) {
+                return std::string("Error parsing JSON:") + err.unparsedArg;
+            }
+        });
+}
 
-    for (unsigned int i = 0; i < n_elem; i++) {
-        members.insert(std::make_pair(vRPCConvertParams[i].methodName,
-                                      vRPCConvertParams[i].paramIdx));
+std::optional<std::pair<std::vector<bool>, std::vector<bool>>>
+ParamsToConvertFor(const std::string& method) {
+    auto search = rpcCvtTable.find(method);
+    if (search != rpcCvtTable.end()) {
+        return search->second;
+    } else {
+        return std::nullopt;
     }
 }
 
-static CRPCConvertTable rpcCvtTable;
-
-/** Non-RFC4627 JSON parser, accepts internal values (such as numbers, true, false, null)
- * as well as objects and arrays.
- */
-UniValue ParseNonRFCJSONValue(const std::string& strVal)
+std::optional<UniValue> ParseNonRFCJSONValue(const std::string& strVal)
 {
     UniValue jVal;
-    if (!jVal.read(std::string("[")+strVal+std::string("]")) ||
-        !jVal.isArray() || jVal.size()!=1)
-        throw runtime_error(string("Error parsing JSON:")+strVal);
-    return jVal[0];
+    if (jVal.read(std::string("[")+strVal+std::string("]")) && jVal.isArray() && jVal.size() == 1) {
+        return jVal[0];
+    } else {
+        return std::nullopt;
+    }
 }
 
-/** Convert strings to command-specific RPC representation */
-UniValue RPCConvertValues(const std::string &strMethod, const std::vector<std::string> &strParams)
+tl::expected<UniValue, ConversionFailure>
+RPCConvertValues(const std::string &method, const std::vector<std::string> &strArgs)
 {
     UniValue params(UniValue::VARR);
+    auto paramsToConvert = ParamsToConvertFor(method);
+    std::vector<bool> requiredParams{}, optionalParams{};
+    if (paramsToConvert.has_value()) {
+        std::tie(requiredParams, optionalParams) = paramsToConvert.value();
+    } else {
+        return tl::expected<UniValue, ConversionFailure>(tl::unexpect, UnknownRPCMethod());
+    }
+
+    if (strArgs.size() < requiredParams.size()
+        || requiredParams.size() + optionalParams.size() < strArgs.size()) {
+        return tl::expected<UniValue, ConversionFailure>(
+                tl::unexpect,
+                WrongNumberOfArguments(requiredParams.size(), optionalParams.size(), strArgs.size()));
+    }
+
+    std::vector<bool> allParams(requiredParams.begin(), requiredParams.end());
+    allParams.reserve(requiredParams.size() + optionalParams.size());
+    allParams.insert(allParams.end(), optionalParams.begin(), optionalParams.end());
 
-    for (unsigned int idx = 0; idx < strParams.size(); idx++) {
-        const std::string& strVal = strParams[idx];
+    for (std::vector<std::string>::size_type idx = 0; idx < strArgs.size(); idx++) {
```

Comment:
Aside: C++20 will apparently have
```c++
    for (const auto& [s, idx] : strArgs | ranges::views::enumerate) { ...
```
There's a library that emulates this: https://github.com/ericniebler/range-v3